### PR TITLE
001 Creating Table

### PR DESCRIPTION
**Pull request by MatthewJKainos**
Added the first migration script to create the tables for the database with referential integrity.

### CODE REVIEW

**Review by Aaron-Millen — APPROVED** [review]
:)

File: src/main/resources/db/migration/V1__creating_tables.sql [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    name varchar(60),
+    role varchar(40),
+    nat_ins_no INT,
+    bank_account varchar(40),
+    salary DECIMAL(10, 2),
+    PRIMARY KEY (employee_id)
+);
+
+CREATE TABLE `Client` (
+    client_id INT AUTO_INCREMENT,
```
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: where is the primary key?

File: src/main/resources/db/migration/V1__creating_tables.sql [hunk 2]
```diff
@@ -0,0 +1,54 @@
+CREATE TABLE Employee (
+    employee_id INT AUTO_INCREMENT,
+    name varchar(60),
+    role varchar(40),
+    nat_ins_no INT,
```
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: whats a nat ins no

File: src/main/resources/db/migration/V1__creating_tables.sql [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    bank_account varchar(40),
+    salary DECIMAL(10, 2),
+    PRIMARY KEY (employee_id)
```
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: are all of thse allowed to be null? What happens if you have a null primary key

File: src/main/resources/db/migration/V1__creating_tables.sql (hunk 2 above)
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: My national insurance number contains letters and isnt 2 million long

File: src/main/resources/db/migration/V1__creating_tables.sql [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+);
+
+CREATE TABLE `Client` (
+    client_id INT AUTO_INCREMENT,
+    name varchar(60),
+    address varchar(120),
```
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: Is this 2nf? Are all address line fields atomic?

File: src/main/resources/db/migration/V1__creating_tables.sql [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    phone_number varchar(20),
+    PRIMARY KEY (client_id)
```
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: can all of these be null?

File: src/main/resources/db/migration/V1__creating_tables.sql [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+);
+
+CREATE TABLE Delivery_Employee_Project (
+    delivery_employee_id INT,
+    project_id INT,
+    working_on_proj BOOLEAN,
+    PRIMARY KEY(delivery_employee_id, working_on_proj)
```
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: can all of these be null?

File: src/main/resources/db/migration/V1__creating_tables.sql [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    name varchar(60),
+    address varchar(120),
+    phone_number varchar(20),
+    PRIMARY KEY (client_id)
+);
+
+CREATE TABLE Delivery_Employee_Project (
+    delivery_employee_id INT,
+    project_id INT,
+    working_on_proj BOOLEAN,
```
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: what is a proj

File: src/main/resources/db/migration/V1__creating_tables.sql [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    PRIMARY KEY(delivery_employee_id, working_on_proj)
+);
+
+CREATE TABLE Technology (
+    tech_id INT,
+    name varchar(60),
+    PRIMARY KEY (tech_id)
```
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: Null?

File: src/main/resources/db/migration/V1__creating_tables.sql [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+CREATE TABLE Delivery_Employee_Project (
+    delivery_employee_id INT,
+    project_id INT,
+    working_on_proj BOOLEAN,
+    PRIMARY KEY(delivery_employee_id, working_on_proj)
+);
+
+CREATE TABLE Technology (
+    tech_id INT,
```
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: whats tech

Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: are we manually defining incrementation?

File: src/main/resources/db/migration/V1__creating_tables.sql [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    name varchar(60),
+    PRIMARY KEY (tech_id)
+);
+
+
+CREATE TABLE Project (
+    project_id INT AUTO_INCREMENT,
```
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: Null?

File: src/main/resources/db/migration/V1__creating_tables.sql [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    start_date DATE,
+    completion_date DATE,
+    PRIMARY KEY (project_id),
+    FOREIGN KEY (tech_lead_id) REFERENCES Employee(employee_id),
+    FOREIGN KEY (sales_employee_id) REFERENCES Employee(employee_id),
+    FOREIGN KEY (client_id) REFERENCES `Client`(client_id)
+);
+
+CREATE TABLE Tech_Proj (
+    tech_id INT,
```
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: Whats tech proj

Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: auto increment? Null?

**Review by lukemKainos — CHANGES_REQUESTED** [review]

File: src/main/resources/db/migration/V1__creating_tables.sql (hunk 1 above)
Comment (MatthewJKainos) [reply src/main/resources/db/migration/V1__creating_tables.sql]: PK constraint is defined at the bottom of the table deffinition

File: src/main/resources/db/migration/V1__creating_tables.sql [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    salary DECIMAL(10, 2) NOT NULL,
+    PRIMARY KEY (employee_id)
+);
+
+CREATE TABLE `Client` (
+    client_id INT AUTO_INCREMENT NOT NULL,
+    name varchar(60) NOT NULL,
+    phone_number varchar(20) NOT NULL,
+    address_line_1 varchar(30) NOT NULL,
+    address_line_2 varchar(30) NOT NULL,
```
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: not all addresses have a line 2

File: src/main/resources/db/migration/V1__creating_tables.sql [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    address_city varchar(20) NOT NULL,
+    address_state varchar(20) NOT NULL,
+    address_country varchar(30) NOT NULL,
+    address_post_code varchar(10) NOT NULL,
```
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: What happens if im from ireland?

File: src/main/resources/db/migration/V1__creating_tables.sql [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+    nation_insurance_no varchar(13) NOT NULL,
+    bank_account varchar(40) NOT NULL,
+    salary DECIMAL(10, 2) NOT NULL,
+    PRIMARY KEY (employee_id)
+);
+
+CREATE TABLE `Client` (
+    client_id INT AUTO_INCREMENT NOT NULL,
+    name varchar(60) NOT NULL,
+    phone_number varchar(20) NOT NULL,
```
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: can you have a phone number of 20?

File: src/main/resources/db/migration/V1__creating_tables.sql [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    address_state varchar(20) NOT NULL,
+    address_country varchar(30) NOT NULL,
+    address_post_code varchar(10) NOT NULL,
+    PRIMARY KEY (client_id)
+);
+
+CREATE TABLE Delivery_Employee_Project (
+    delivery_employee_id INT NOT NULL,
+    project_id INT NOT NULL,
+    working_on_proj BOOLEAN NOT NULL,
```
Comment (lukemKainos) [inline src/main/resources/db/migration/V1__creating_tables.sql]: proj?

**Review by lukemKainos — CHANGES_REQUESTED** [review]

**Review by lukemKainos — APPROVED** [review]

**Pull request merged**